Firestore/core/test/unit/model/values_test.cc

```diff
@@ -0,0 +1,323 @@
+  Add(comparison_groups, Wrap("a"));
+  Add(comparison_groups, Wrap("abc def"));
+  // latin small letter e + combining acute accent + latin small letter b
+  Add(comparison_groups, Wrap("e\u0301b"));
+  Add(comparison_groups, Wrap("æ"));
+  // latin small letter e with acute accent + latin small letter a
+  Add(comparison_groups, Wrap("\u00e9a"));
+
+  // blobs
+  Add(comparison_groups, Wrap(BlobValue()));
+  Add(comparison_groups, Wrap(BlobValue(0)));
+  Add(comparison_groups, Wrap(BlobValue(0, 1, 2, 3, 4)));
+  Add(comparison_groups, Wrap(BlobValue(0, 1, 2, 4, 3)));
+  Add(comparison_groups, Wrap(BlobValue(255)));
+
+  // resource names
+  Add(comparison_groups, WrapReference(DbId("p1/d1"), Key("c1/doc1")));
+  Add(comparison_groups, WrapReference(DbId("p1/d1"), Key("c1/doc2")));
+  Add(comparison_groups, WrapReference(DbId("p1/d1"), Key("c10/doc1")));
+  Add(comparison_groups, WrapReference(DbId("p1/d1"), Key("c2/doc1")));
+  Add(comparison_groups, WrapReference(DbId("p1/d2"), Key("c1/doc1")));
+  Add(comparison_groups, WrapReference(DbId("p2/d1"), Key("c1/doc1")));
+
+  // geo points
+  Add(comparison_groups, Wrap(GeoPoint(-90, -180)));
+  Add(comparison_groups, Wrap(GeoPoint(-90, 0)));
+  Add(comparison_groups, Wrap(GeoPoint(-90, 180)));
+  Add(comparison_groups, Wrap(GeoPoint(0, -180)));
+  Add(comparison_groups, Wrap(GeoPoint(0, 0)));
+  Add(comparison_groups, Wrap(GeoPoint(0, 180)));
+  Add(comparison_groups, Wrap(GeoPoint(1, -180)));
+  Add(comparison_groups, Wrap(GeoPoint(1, 0)));
+  Add(comparison_groups, Wrap(GeoPoint(1, 180)));
+  Add(comparison_groups, Wrap(GeoPoint(90, -180)));
+  Add(comparison_groups, Wrap(GeoPoint(90, 0)));
+  Add(comparison_groups, Wrap(GeoPoint(90, 180)));
+
+  // arrays
+  Add(comparison_groups, WrapArray("bar"));
+  Add(comparison_groups, WrapArray("foo", 1));
+  Add(comparison_groups, WrapArray("foo", 2));
+  Add(comparison_groups, WrapArray("foo", "0"));
+
+  // objects
+  Add(comparison_groups, WrapObject("bar", 0));
+  Add(comparison_groups, WrapObject("bar", 0, "foo", 1));
+  Add(comparison_groups, WrapObject("foo", 1));
+  Add(comparison_groups, WrapObject("foo", 2));
+  Add(comparison_groups, WrapObject("foo", "0"));
+
+  for (size_t i = 0; i < comparison_groups.size(); ++i) {
+    for (size_t j = i; j < comparison_groups.size(); ++j) {
+      if (i == j) {
+        VerifyOrdering(comparison_groups[i], comparison_groups[i],
+                       ComparisonResult::Same);
+      } else {
+        VerifyOrdering(comparison_groups[i], comparison_groups[j],
+                       ComparisonResult::Ascending);
+      }
+    }
+  }
+}
+
+TEST_F(ValuesTest, CanonicalId) {
+  VerifyCanonicalId(Wrap(nullptr), "null");
+  VerifyCanonicalId(Wrap(true), "true");
+  VerifyCanonicalId(Wrap(false), "false");
+  VerifyCanonicalId(Wrap(1), "1");
+  VerifyCanonicalId(Wrap(1.0), "1.000000");
+  VerifyCanonicalId(Wrap(Timestamp(30, 1000)), "time(30,1000)");
+  VerifyCanonicalId(Wrap("a"), "a");
```

Comment:
"a\0b" should also work here.

Note you can create above with `std::string("a\0b", 3)`, spelling the length `3` is required.